addons/radar/functions/fnc_cacheLoop.sqf

```diff
@@ -1,263 +1,263 @@
-            _curList ctrlCommit 0;
-            _text = "";
-        };
-        if (count _lists >= (_listIndex + 1)) then {
-            _curList = _lists # _listIndex;
-        } else {
-            ctrlPosition _grpCtrl params ["_left", "_top", "_width", "_height"];
-            // create group
-            private _curGrp = _display ctrlCreate["RscControlsGroupNoScrollbars", -1, _grpCtrl];
-            private _ctrlPos = [
-                (0 * pixelW) * _listIndex + _listWidth * _listIndex,
-                0,
-                _listWidth,
-                _listHeight
-            ];
-            _curGrp ctrlSetPosition _ctrlPos;
-            _curGrp ctrlCommit 0;
-
-            _curList = _display ctrlCreate ["RscStructuredText", -1, _curGrp];
-            _curList ctrlSetFont diwako_dui_font;
-            _curList ctrlSetBackgroundColor [0,0,0,_bgOpacity];
-            _lists pushBack _curList;
-            _curList ctrlCommit 0;
-        };
-        _listIndex = _listIndex + 1;
-    };
-    private _unit = _x;
-    private _selected = "";
-    if ((count _selectedUnits) > 0 && {_unit != _player}) then {
-        private _curName = vehicleVarName _unit;
-        _unit setVehicleVarName "";
-        private _defaultIdent = str _unit;
-        _unit setVehicleVarName _curName;
-        private _arr = [_defaultIdent, ":"] call CBA_fnc_split;
-        private _num = if ((count _arr) > 1) then {
-            (_arr select 1) select [0, 2]
-        } else {
-            ""
-        };
-        _selected = format ["%1%2", (["", ">> "] select (_selectedUnits findIf {_x == _unit} > -1)), _num];
-    };
-    private _buddy = ["", _iconNamespace getVariable ["buddy", DUI_BUDDY]] select (_player == (_unit getVariable ["diwako_dui_buddy", objNull]));
-    private _icon = [_unit getVariable ["diwako_dui_icon", DUI_RIFLEMAN], ""] select (_buddy != "" && {_onlyBuddyIcon});
-    _text = format ["%1<t color='%4' size='%6' shadow='%8' shadowColor='#000000' valign='middle' align='left'>%5<img image='%7'valign='bottom'/><img image='%2'valign='bottom'/> %3</t><br/>",
-        _text, // 1
-        _icon, // 2
-        _unit getVariable ["ACE_Name", name _unit], // 3
-        _unit getVariable ["diwako_dui_color","#FFFFFF"], // 4
-        _selected, // 5
-        (_textSize * _heightMod), // 6
-        _buddy, // 7
-        _shadow]; // 8
-} forEach _group;
-
-if !(isNull _curList) then {
-    _curList ctrlSetStructuredText parseText _text;
-    _curList ctrlSetPosition _ctrlPosList;
-    _curList ctrlCommit 0;
-};
-for "_i" from (count _lists) -1 to _listIndex step -1 do {
-    ctrlDelete ctrlParentControlsGroup (_lists deleteAt _i);
-};
+#include "script_component.hpp"
+
+// loop
+[FUNC(cacheLoop),[],0.5] call CBA_fnc_waitAndExecute;
+
+// if both compass and namelist are not enabled, just remove the controls if there are any
+if !(diwako_dui_enable_compass || diwako_dui_namelist) exitWith {
+    for "_i" from 0 to (count diwako_dui_namebox_lists) do {
```

Comment:
Why not forEach, and then set to empty array at once.

Reply:
That I can answer, I tried forEach, but if you remove an element forEach Will stop prematurely leaving a few elements in there.
In the end I stopped thinking of how to write it better as it worked out plus when both namelist and compass are disabled it is not setting any variables.

Reply:
`but if you remove an element forEach` Jup. Because it iterates by index.
But you can just iterate first, and then delete all elements. Instead of iterating each element and reshuffling the whole array X times.